File: uikit-custom-sample/src/main/java/com/sendbird/uikit/customsample/SettingsFragment.java
package com.sendbird.uikit.customsample;

import static android.app.Activity.RESULT_OK;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.sendbird.android.SendbirdChat;
import com.sendbird.android.params.UserUpdateParams;
import com.sendbird.android.user.User;
import com.sendbird.uikit.SendbirdUIKit;
import com.sendbird.uikit.consts.DialogEditTextParams;
import com.sendbird.uikit.customsample.consts.StringSet;
import com.sendbird.uikit.customsample.utils.DrawableUtils;
import com.sendbird.uikit.customsample.utils.PreferenceUtils;
import com.sendbird.uikit.interfaces.OnEditTextResultListener;
import com.sendbird.uikit.log.Logger;
import com.sendbird.uikit.model.DialogListItem;
import com.sendbird.uikit.utils.ContextUtils;
import com.sendbird.uikit.utils.DialogUtils;
import com.sendbird.uikit.utils.FileUtils;
import com.sendbird.uikit.utils.IntentUtils;
import com.sendbird.uikit.utils.PermissionUtils;
import com.sendbird.uikit.utils.TextUtils;
import com.sendbird.uikit.widgets.WaitingDialog;

import java.io.File;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Displays a settings screen.
 */
public class SettingsFragment extends Fragment {
    private final String[] REQUIRED_PERMISSIONS = PermissionUtils.CAMERA_PERMISSION;

    private AppCompatImageView profileImageView;
    private TextView nicknameTextView;
    private SwitchCompat disturbSwitch;
    private Uri mediaUri;

    private final ActivityResultLauncher<Intent> appSettingLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), intent -> {
        if (getContext() == null) return;
        final boolean hasPermission = PermissionUtils.hasPermissions(getContext(), REQUIRED_PERMISSIONS);
        if (hasPermission) {
            showMediaSelectDialog();
        }
    });
    private final ActivityResultLauncher<String[]> permissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), permissionResults -> {
        if (PermissionUtils.getNotGrantedPermissions(permissionResults).isEmpty()) {
            showMediaSelectDialog();
        }
    });

    private final ActivityResultLauncher<Intent> takeCameraLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
        SendbirdChat.setAutoBackgroundDetection(true);
        int resultCode = result.getResultCode();

        if (resultCode != RESULT_OK || getContext() == null) return;
        final Uri mediaUri = this.mediaUri;
        if (mediaUri != null) {
            final File file = FileUtils.uriToFile(getContext().getApplicationContext(), mediaUri);
            updateUserProfileImage(file);
        }
    });
    private final ActivityResultLauncher<Intent> getContentLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
        SendbirdChat.setAutoBackgroundDetection(true);
        final Intent intent = result.getData();
        int resultCode = result.getResultCode();

        if (resultCode != RESULT_OK || intent == null || getContext() == null) return;
        final Uri mediaUri = intent.getData();
        if (mediaUri != null) {
            final File file = FileUtils.uriToFile(getContext().getApplicationContext(), mediaUri);
            updateUserProfileImage(file);
        }
    });

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        SendbirdUIKit.connect(null);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_settings, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initPage(view);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        SendbirdChat.setAutoBackgroundDetection(true);
    }

    private void initPage(@NonNull View view) {
        if (getContext() == null) {
            return;
        }

        Toolbar toolbar = view.findViewById(R.id.tbSettings);
        toolbar.inflateMenu(R.menu.settings_menu);
        toolbar.getMenu().findItem(R.id.action_edit_profile)
                .getActionView().setOnClickListener(v -> {
            Logger.d("++ edit button clicked");
            showEditProfileDialog();
        });

        boolean useDoNotDisturb = true;
        if (getArguments() != null) {
            useDoNotDisturb = getArguments().getBoolean(StringSet.SETTINGS_USE_DO_NOT_DISTURB, true);
        }

        profileImageView = view.findViewById(R.id.ivProfileView);
        loadUserProfileUrl(PreferenceUtils.getProfileUrl());
        TextView userIdTextView = view.findViewById(R.id.tvUserId);
        userIdTextView.setText(PreferenceUtils.getUserId());
        nicknameTextView = view.findViewById(R.id.tvNickname);
        nicknameTextView.setText(PreferenceUtils.getNickname());

        View disturbItem = view.findViewById(R.id.itemDisturb);
        disturbItem.setVisibility(useDoNotDisturb ? View.VISIBLE : View.GONE);
        if (useDoNotDisturb) {
            disturbItem.setOnClickListener(v -> {
                Logger.d("++ disturb clicked");
                updateDoNotDisturb();
            });

            disturbSwitch = view.findViewById(R.id.scDisturbSwitch);
            SendbirdChat.getDoNotDisturb((b, i, i1, i2, i3, s, e) -> {
                PreferenceUtils.setDoNotDisturb(b);
                disturbSwitch.setChecked(PreferenceUtils.getDoNotDisturb());
            });
            disturbSwitch.setOnClickListener(v -> {
                Logger.d("++ disturb clicked");
                updateDoNotDisturb();
            });
        }

        View signOutItem = view.findViewById(R.id.itemHome);
        signOutItem.setOnClickListener(v -> {
            Logger.d("++ home clicked");
            if (getActivity() != null) {
                getActivity().finish();
            }
        });
    }

    private void showEditProfileDialog() {
        if (getContext() == null || getFragmentManager() == null) return;

        DialogListItem[] items = {
                new DialogListItem(R.string.text_settings_change_user_nickname),
                new DialogListItem(R.string.text_settings_change_user_profile_image)
        };

        DialogUtils.showListBottomDialog(requireContext(), items, (v, p, item) -> {
            final int key = item.getKey();
            if (key == R.string.text_settings_change_user_nickname) {
                Logger.dev("change user nickname");
                OnEditTextResultListener listener = result -> {
                    if (TextUtils.isEmpty(result)) {
                        return;
                    }
                    updateUserNickname(result);
                };

                DialogEditTextParams params = new DialogEditTextParams(getString(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_name_hint));
                params.setEnableSingleLine(true);
                DialogUtils.showInputDialog(
                        requireContext(),
                        getString(R.string.text_settings_change_user_nickname),
                        params, listener,
                        getString(com.sendbird.uikit.R.string.sb_text_button_save), null,
                        getString(com.sendbird.uikit.R.string.sb_text_button_cancel), null);
            } else if (key == R.string.text_settings_change_user_profile_image) {
                Logger.dev("change user profile");

                if (getContext() == null) return;
                final boolean hasPermission = PermissionUtils.hasPermissions(getContext(), REQUIRED_PERMISSIONS);
                if (hasPermission) {
                    showMediaSelectDialog();
                    return;
                }

                requestPermission(REQUIRED_PERMISSIONS);
            }
        });
    }

    private void requestPermission(@NonNull String[] permissions) {
        if (getContext() == null || getActivity() == null) return;
        // 1. check permission
        final boolean hasPermission = PermissionUtils.hasPermissions(getContext(), permissions);
        if (hasPermission) {
            showMediaSelectDialog();
            return;
        }

        // 2. determine whether rationale popup should show
        final List<String> deniedList = PermissionUtils.getExplicitDeniedPermissionList(getActivity(), permissions);
        if (!deniedList.isEmpty()) {
            showPermissionRationalePopup(deniedList.get(0));
            return;
        }
        // 3. request permission
        this.permissionLauncher.launch(permissions);
    }

    private void showPermissionRationalePopup(@NonNull String permission) {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireContext());
        builder.setTitle(getString(com.sendbird.uikit.R.string.sb_text_dialog_permission_title));
        builder.setMessage(getPermissionGuideMessage(requireContext(), permission));
        builder.setPositiveButton(com.sendbird.uikit.R.string.sb_text_go_to_settings, (dialogInterface, i) -> {
            Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            intent.addCategory(Intent.CATEGORY_DEFAULT);
            intent.setData(Uri.parse("package:" + requireContext().getPackageName()));
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
            intent.addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
            appSettingLauncher.launch(intent);
        });
        AlertDialog dialog = builder.create();
        dialog.show();
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(ContextCompat.getColor(requireContext(), com.sendbird.uikit.R.color.secondary_300));
    }

    private void updateUserNickname(@NonNull String nickname) {
        if (getContext() == null) return;
        WaitingDialog.show(getContext());
        final UserUpdateParams params = new UserUpdateParams();
        params.setNickname(nickname);
        SendbirdUIKit.updateUserInfo(params, e -> {
            WaitingDialog.dismiss();
            if (e != null) {
                Logger.e(e);
                return;
            }

            PreferenceUtils.setNickname(nickname);
            nicknameTextView.setText(nickname);
        });
    }

    private void updateUserProfileImage(@NonNull File profileImage) {
        if (getContext() == null) return;
        WaitingDialog.show(getContext());
        final UserUpdateParams params = new UserUpdateParams();
        params.setProfileImageFile(profileImage);
        SendbirdUIKit.updateUserInfo(params, e -> {
            WaitingDialog.dismiss();
            if (e != null) {
                Logger.e(e);
                return;
            }

            final User currentUser = SendbirdChat.getCurrentUser();
            if (currentUser != null) {
                final String profileUrl = currentUser.getProfileUrl();
                PreferenceUtils.setProfileUrl(profileUrl);
                loadUserProfileUrl(profileUrl);
            }
        });
    }

    private void updateDoNotDisturb() {
        disturbSwitch.setChecked(!PreferenceUtils.getDoNotDisturb());
        Logger.d("update do not disturb : " + !PreferenceUtils.getDoNotDisturb());
        SendbirdChat.setDoNotDisturb(!PreferenceUtils.getDoNotDisturb(), 0, 0, 23, 59, TimeZone.getDefault().getID(), e -> {
            if (e != null) {
                disturbSwitch.setChecked(PreferenceUtils.getDoNotDisturb());
                return;
            }
            Logger.d("update do not disturb on callback : " + !PreferenceUtils.getDoNotDisturb());
            PreferenceUtils.setDoNotDisturb(!PreferenceUtils.getDoNotDisturb());
        });
    }

    private void showMediaSelectDialog() {
        if (getContext() == null) return;
        DialogListItem[] items = {
                new DialogListItem(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_camera),
                new DialogListItem(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_gallery)};

        DialogUtils.showListDialog(requireContext(),
                getString(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image),
                items, (v, p, item) -> {
                    try {
                        final int key = item.getKey();
                        if (key == com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_camera) {
                            takeCamera();
                        } else if (key == com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_gallery) {
                            takePhoto();
                        }
                    } catch (Exception e) {
                        Logger.e(e);
                    }
                });
    }

    private void takeCamera() {
        SendbirdChat.setAutoBackgroundDetection(false);
        this.mediaUri = FileUtils.createPictureImageUri(requireContext());
        if (mediaUri == null) return;
        Intent intent = IntentUtils.getCameraIntent(requireContext(), mediaUri);
        if (IntentUtils.hasIntent(requireContext(), intent)) {
            takeCameraLauncher.launch(intent);
        }
    }

    private void takePhoto() {
        SendbirdChat.setAutoBackgroundDetection(false);
        Intent intent = IntentUtils.getImageGalleryIntent();
        getContentLauncher.launch(intent);
    }

    private static String getPermissionGuideMessage(@NonNull Context context, @NonNull String permission) {
        int textResId;
        if (Manifest.permission.CAMERA.equals(permission)) {
            textResId = com.sendbird.uikit.R.string.sb_text_need_to_allow_permission_camera;
        } else {
            textResId = com.sendbird.uikit.R.string.sb_text_need_to_allow_permission_storage;
        }
        return String.format(Locale.US, context.getString(textResId), ContextUtils.getApplicationName(context));
    }

    private void loadUserProfileUrl(final String url) {
        if (getContext() != null) {
            Glide.with(requireContext())
                    .load(url)
                    .circleCrop()
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .error(DrawableUtils.setTintList(requireContext(), R.drawable.icon_user, SendbirdUIKit.getDefaultThemeMode().getMonoTintResId()))
                    .into(profileImageView);
        }
    }
}

File: uikit-sample/src/main/kotlin/com/sendbird/uikit_messaging_android/SettingsFragment.kt
package com.sendbird.uikit_messaging_android

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.sendbird.android.SendbirdChat
import com.sendbird.android.SendbirdChat.autoBackgroundDetection
import com.sendbird.android.SendbirdChat.currentUser
import com.sendbird.android.SendbirdChat.getDoNotDisturb
import com.sendbird.android.SendbirdChat.setDoNotDisturb
import com.sendbird.android.SendbirdChat.updateCurrentUserInfo
import com.sendbird.android.exception.SendbirdException
import com.sendbird.android.handler.CompletionHandler
import com.sendbird.android.handler.DoNotDisturbHandler
import com.sendbird.android.params.UserUpdateParams
import com.sendbird.uikit.SendbirdUIKit
import com.sendbird.uikit.consts.DialogEditTextParams
import com.sendbird.uikit.interfaces.OnEditTextResultListener
import com.sendbird.uikit.log.Logger
import com.sendbird.uikit.model.DialogListItem
import com.sendbird.uikit.modules.components.StateHeaderComponent
import com.sendbird.uikit.utils.ContextUtils
import com.sendbird.uikit.utils.DialogUtils
import com.sendbird.uikit.utils.FileUtils
import com.sendbird.uikit.utils.IntentUtils
import com.sendbird.uikit.utils.PermissionUtils
import com.sendbird.uikit.utils.TextUtils
import com.sendbird.uikit_messaging_android.consts.StringSet
import com.sendbird.uikit_messaging_android.databinding.FragmentSettingsBinding
import com.sendbird.uikit_messaging_android.databinding.ViewCustomMenuTextButtonBinding
import com.sendbird.uikit_messaging_android.utils.DrawableUtils.setTintList
import com.sendbird.uikit_messaging_android.utils.PreferenceUtils
import com.sendbird.uikit_messaging_android.widgets.WaitingDialog
import java.io.File
import java.util.Locale
import java.util.TimeZone

/**
 * Displays a settings screen.
 */
class SettingsFragment : Fragment() {
    private lateinit var binding: FragmentSettingsBinding
    private val headerComponent = StateHeaderComponent()
    private var mediaUri: Uri? = null
    private val appSettingLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { _: ActivityResult? ->
            if (context == null) return@registerForActivityResult
            val hasPermission = PermissionUtils.hasPermissions(requireContext(), *REQUIRED_PERMISSIONS)
            if (hasPermission) {
                showMediaSelectDialog()
            }
        }
    private val permissionLauncher =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) {
            if (PermissionUtils.getNotGrantedPermissions(it).isEmpty()) {
                showMediaSelectDialog()
            }
        }
    private val takeCameraLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
            autoBackgroundDetection = true
            val resultCode = result.resultCode
            if (resultCode != Activity.RESULT_OK || context == null) return@registerForActivityResult
            val mediaUri = mediaUri
            if (mediaUri != null) {
                val file = FileUtils.uriToFile(requireContext().applicationContext, mediaUri)
                updateUserProfileImage(file)
            }
        }
    private val getContentLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
            autoBackgroundDetection = true
            val intent = result.data
            val resultCode = result.resultCode
            if (resultCode != Activity.RESULT_OK || intent == null || context == null) return@registerForActivityResult
            val mediaUri = intent.data
            if (mediaUri != null) {
                val file = FileUtils.uriToFile(requireContext().applicationContext, mediaUri)
                updateUserProfileImage(file)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        SendbirdUIKit.connect(null)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentSettingsBinding.inflate(inflater, container, false)
        headerComponent.params.title = getString(R.string.text_settings_header_title)
        headerComponent.params.setUseLeftButton(false)
        headerComponent.params.rightButtonText = getString(R.string.text_settings_header_edit_button)
        val header =
            headerComponent.onCreateView(requireContext(), inflater, binding.headerComponent, savedInstanceState)
        binding.headerComponent.addView(header)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initPage()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.settings_menu, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        val editMenuItem = menu.findItem(R.id.action_edit_profile)
        val binding = ViewCustomMenuTextButtonBinding.inflate(
            layoutInflater
        )
        val isDark = PreferenceUtils.isUsingDarkTheme
        binding.text.setText(R.string.text_settings_header_edit_button)
        binding.text.setBackgroundResource(if (isDark) R.drawable.sb_button_uncontained_background_dark else R.drawable.sb_button_uncontained_background_light)
        binding.text.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.primary_200 else R.color.primary_300,
                null
            )
        )
        val rootView: View = binding.root
        rootView.setOnClickListener { onOptionsItemSelected(editMenuItem) }
        editMenuItem.actionView = rootView
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_edit_profile) {
            Logger.d("++ edit button clicked")
            showEditProfileDialog()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        setHasOptionsMenu(true)
    }

    override fun onDestroy() {
        super.onDestroy()
        autoBackgroundDetection = true
    }

    private fun requestPermission(permissions: Array<String>) {
        if (context == null || activity == null) return
        // 1. check permission
        val hasPermission = PermissionUtils.hasPermissions(requireContext(), *permissions)
        if (hasPermission) {
            showMediaSelectDialog()
            return
        }

        // 2. determine whether rationale popup should show
        val deniedList = PermissionUtils.getExplicitDeniedPermissionList(requireActivity(), *permissions)
        if (deniedList.isNotEmpty()) {
            showPermissionRationalePopup(deniedList[0])
            return
        }
        // 3. request permission
        permissionLauncher.launch(permissions)
    }

    private fun showPermissionRationalePopup(permission: String) {
        val builder = AlertDialog.Builder(requireContext())
        builder.setTitle(getString(com.sendbird.uikit.R.string.sb_text_dialog_permission_title))
        builder.setMessage(getPermissionGuideMessage(requireContext(), permission))
        builder.setPositiveButton(com.sendbird.uikit.R.string.sb_text_go_to_settings) { _: DialogInterface?, _: Int ->
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
            intent.addCategory(Intent.CATEGORY_DEFAULT)
            intent.data = Uri.parse("package:" + requireContext().packageName)
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY)
            intent.addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS)
            appSettingLauncher.launch(intent)
        }
        val dialog = builder.create()
        dialog.show()
        dialog.getButton(AlertDialog.BUTTON_POSITIVE)
            .setTextColor(ContextCompat.getColor(requireContext(), com.sendbird.uikit.R.color.secondary_300))
    }

    private fun initPage() {
        if (context == null) {
            return
        }
        val isDark = PreferenceUtils.isUsingDarkTheme
        binding.background.setBackgroundResource(if (isDark) R.color.background_600 else R.color.background_50)
        binding.tvNickname.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_01 else R.color.onlight_01,
                null
            )
        )
        binding.idDivider.setBackgroundResource(if (isDark) R.drawable.sb_line_divider_dark else R.drawable.sb_line_divider_light)
        binding.tvUserIdBadge.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_02 else R.color.onlight_02,
                null
            )
        )
        binding.tvUserId.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_01 else R.color.onlight_01,
                null
            )
        )
        binding.profileDivider.setBackgroundResource(if (isDark) R.drawable.sb_line_divider_dark else R.drawable.sb_line_divider_light)
        binding.itemDarkTheme.setBackgroundResource(if (isDark) R.drawable.selector_rectangle_dark600 else R.drawable.selector_rectangle_light)
        binding.tvDarkThemeName.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_01 else R.color.onlight_01,
                null
            )
        )
        binding.darkThemeDivider.setBackgroundResource(if (isDark) R.drawable.sb_line_divider_dark else R.drawable.sb_line_divider_light)
        binding.itemDisturb.setBackgroundResource(if (isDark) R.drawable.selector_rectangle_dark600 else R.drawable.selector_rectangle_light)
        binding.tvDisturbName.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_01 else R.color.onlight_01,
                null
            )
        )
        binding.disturbDivider.setBackgroundResource(if (isDark) R.drawable.sb_line_divider_dark else R.drawable.sb_line_divider_light)
        binding.itemHome.setBackgroundResource(if (isDark) R.drawable.selector_rectangle_dark600 else R.drawable.selector_rectangle_light)
        binding.tvHomeName.setTextColor(
            ResourcesCompat.getColor(
                resources,
                if (isDark) R.color.ondark_01 else R.color.onlight_01,
                null
            )
        )
        binding.signOutDivider.setBackgroundResource(if (isDark) R.drawable.sb_line_divider_dark else R.drawable.sb_line_divider_light)
        val useHeader = arguments?.getBoolean(StringSet.SETTINGS_USE_HEADER, true) ?: true
        val useDoNotDisturb = arguments?.getBoolean(StringSet.SETTINGS_USE_DO_NOT_DISTURB, true) ?: true
        headerComponent.rootView?.visibility = if (useHeader) View.VISIBLE else View.GONE
        headerComponent.setOnRightButtonClickListener {
            Logger.d("++ edit button clicked")
            showEditProfileDialog()
        }
        val currentUser = SendbirdChat.currentUser
        if (currentUser != null) {
            loadUserProfileUrl(currentUser.profileUrl)
            binding.tvUserId.text = currentUser.userId
            binding.tvNickname.text = currentUser.nickname
        } else {
            loadUserProfileUrl(PreferenceUtils.profileUrl)
            binding.tvUserId.text = PreferenceUtils.userId
            binding.tvNickname.text = PreferenceUtils.nickname
        }
        val isDarkMode = PreferenceUtils.isUsingDarkTheme
        val switchTrackTint =
            if (isDarkMode) com.sendbird.uikit.R.color.sb_switch_track_dark else com.sendbird.uikit.R.color.sb_switch_track_light
        val switchThumbTint =
            if (isDarkMode) com.sendbird.uikit.R.color.sb_switch_thumb_dark else com.sendbird.uikit.R.color.sb_switch_thumb_light
        binding.itemDarkTheme.setOnClickListener {
            Logger.d("++ dark theme clicked")
            updateDarkTheme()
        }
        val iconTint = if (SendbirdUIKit.isDarkMode()) R.color.background_700 else R.color.background_50
        val themeBackgroundTint = if (SendbirdUIKit.isDarkMode()) R.color.background_300 else R.color.background_400
        binding.ivDarkThemeIcon.setImageDrawable(
            setTintList(
                requireContext(), R.drawable.icon_theme, iconTint
            )
        )
        binding.ivDarkThemeIcon.setBackgroundDrawable(
            setTintList(
                requireContext(), R.drawable.shape_oval, themeBackgroundTint
            )
        )
        binding.scDarkThemeSwitch.trackTintList = AppCompatResources.getColorStateList(
            requireContext(), switchTrackTint
        )
        binding.scDarkThemeSwitch.thumbTintList = AppCompatResources.getColorStateList(
            requireContext(), switchThumbTint
        )
        binding.scDarkThemeSwitch.isChecked = PreferenceUtils.isUsingDarkTheme
        SendbirdUIKit.setDefaultThemeMode(if (PreferenceUtils.isUsingDarkTheme) SendbirdUIKit.ThemeMode.Dark else SendbirdUIKit.ThemeMode.Light)
        binding.scDarkThemeSwitch.setOnClickListener {
            Logger.d("++ dark theme clicked")
            updateDarkTheme()
        }
        val disturbBackgroundTint = if (SendbirdUIKit.isDarkMode()) R.color.secondary_200 else R.color.secondary_300
        binding.ivDisturbIcon.setImageDrawable(
            setTintList(
                requireContext(), R.drawable.icon_notifications_filled, iconTint
            )
        )
        binding.ivDisturbIcon.setBackgroundDrawable(
            setTintList(
                requireContext(), R.drawable.shape_oval, disturbBackgroundTint
            )
        )
        binding.itemDisturb.visibility = if (useDoNotDisturb) View.VISIBLE else View.GONE
        if (useDoNotDisturb) {
            binding.itemDisturb.setOnClickListener {
                Logger.d("++ disturb clicked")
                updateDoNotDisturb()
            }
            binding.scDisturbSwitch.trackTintList = AppCompatResources.getColorStateList(
                requireContext(), switchTrackTint
            )
            binding.scDisturbSwitch.thumbTintList = AppCompatResources.getColorStateList(
                requireContext(), switchThumbTint
            )
            getDoNotDisturb(DoNotDisturbHandler { b: Boolean, _: Int, _: Int, _: Int, _: Int, _: String?, _: SendbirdException? ->
                PreferenceUtils.doNotDisturb = b
                if (isActive) {
                    binding.scDisturbSwitch.isChecked = PreferenceUtils.doNotDisturb
                }
            })
            binding.scDisturbSwitch.setOnClickListener {
                Logger.d("++ disturb clicked")
                updateDoNotDisturb()
            }
        }
        val homeBackgroundTint = SendbirdUIKit.getDefaultThemeMode().errorColorResId
        binding.ivHomeIcon.setImageDrawable(
            setTintList(
                requireContext(), R.drawable.icon_leave, iconTint
            )
        )
        binding.ivHomeIcon.setBackgroundDrawable(
            setTintList(
                requireContext(), R.drawable.shape_oval, homeBackgroundTint
            )
        )
        binding.itemHome.setOnClickListener {
            Logger.d("++ home clicked")
            activity?.finish()
        }
    }

    private fun showEditProfileDialog() {
        if (context == null) return
        val items = arrayOf(
            DialogListItem(R.string.text_settings_change_user_nickname),
            DialogListItem(R.string.text_settings_change_user_profile_image)
        )
        DialogUtils.showListBottomDialog(requireContext(), items) { _: View?, _: Int, item: DialogListItem ->
            val key = item.key
            if (key == R.string.text_settings_change_user_nickname) {
                Logger.dev("change user nickname")
                val listener = OnEditTextResultListener { result: String ->
                    if (TextUtils.isEmpty(result)) {
                        return@OnEditTextResultListener
                    }
                    updateUserNickname(result)
                }
                val params = DialogEditTextParams(getString(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_name_hint))
                params.enableSingleLine = true
                DialogUtils.showInputDialog(
                    requireContext(),
                    getString(R.string.text_settings_change_user_nickname),
                    params, listener,
                    getString(com.sendbird.uikit.R.string.sb_text_button_save), null,
                    getString(com.sendbird.uikit.R.string.sb_text_button_cancel), null
                )
            } else if (key == R.string.text_settings_change_user_profile_image) {
                Logger.dev("change user profile")
                if (context == null) return@showListBottomDialog
                val hasPermission = PermissionUtils.hasPermissions(requireContext(), *REQUIRED_PERMISSIONS)
                if (hasPermission) {
                    showMediaSelectDialog()
                    return@showListBottomDialog
                }
                requestPermission(REQUIRED_PERMISSIONS)
            }
        }
    }

    private fun updateUserNickname(nickname: String) {
        if (context == null) return
        WaitingDialog.show(requireContext())
        val params = UserUpdateParams()
        params.nickname = nickname
        SendbirdUIKit.updateUserInfo(params) { e: SendbirdException? ->
            WaitingDialog.dismiss()
            if (e != null) {
                Logger.e(e)
                ContextUtils.toastError(context, R.string.text_error_update_user_information)
                return@updateUserInfo
            }
            PreferenceUtils.nickname = nickname
            if (isActive) {
                binding.tvNickname.text = nickname
            }
        }
    }

    private fun updateUserProfileImage(profileImage: File) {
        if (context == null) return
        WaitingDialog.show(requireContext())
        val params = UserUpdateParams()
        params.profileImageFile = profileImage
        updateCurrentUserInfo(params, CompletionHandler { e: SendbirdException? ->
            WaitingDialog.dismiss()
            if (e != null) {
                Logger.e(e)
                ContextUtils.toastError(context, R.string.text_error_update_user_information)
                return@CompletionHandler
            }
            val currentUser = currentUser
            if (currentUser != null) {
                val profileUrl = currentUser.profileUrl
                PreferenceUtils.profileUrl = profileUrl
                loadUserProfileUrl(profileUrl)
            }
        })
    }

    private fun updateDarkTheme() {
        SendbirdUIKit.setDefaultThemeMode(if (SendbirdUIKit.isDarkMode()) SendbirdUIKit.ThemeMode.Light else SendbirdUIKit.ThemeMode.Dark)
        PreferenceUtils.isUsingDarkTheme = SendbirdUIKit.isDarkMode()
        binding.scDarkThemeSwitch.isChecked = SendbirdUIKit.isDarkMode()
        if (activity != null) {
            requireActivity().finish()
            startActivity(requireActivity().intent)
        }
    }

    private fun updateDoNotDisturb() {
        binding.scDisturbSwitch.isChecked = !PreferenceUtils.doNotDisturb
        Logger.d("update do not disturb : " + !PreferenceUtils.doNotDisturb)
        setDoNotDisturb(
            !PreferenceUtils.doNotDisturb,
            0,
            0,
            23,
            59,
            TimeZone.getDefault().id,
            CompletionHandler { e: SendbirdException? ->
                if (e != null) {
                    ContextUtils.toastError(context, R.string.text_error_update_do_not_disturb)
                    if (isActive) {
                        binding.scDisturbSwitch.isChecked = PreferenceUtils.doNotDisturb
                    }
                    return@CompletionHandler
                }
                Logger.d("update do not disturb on callback : " + !PreferenceUtils.doNotDisturb)
                PreferenceUtils.doNotDisturb = !PreferenceUtils.doNotDisturb
            })
    }

    private fun showMediaSelectDialog() {
        if (context == null) return
        val items = arrayOf(
            DialogListItem(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_camera),
            DialogListItem(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_gallery)
        )
        DialogUtils.showListDialog(
            requireContext(),
            getString(com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image),
            items
        ) { _: View?, _: Int, item: DialogListItem ->
            try {
                val key = item.key
                if (key == com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_camera) {
                    takeCamera()
                } else if (key == com.sendbird.uikit.R.string.sb_text_channel_settings_change_channel_image_gallery) {
                    takePhoto()
                }
            } catch (e: Exception) {
                Logger.e(e)
            }
        }
    }

    private fun takeCamera() {
        autoBackgroundDetection = false
        mediaUri = FileUtils.createImageFileUri(requireContext())
        val uri = mediaUri ?: return
        val intent = IntentUtils.getCameraIntent(requireContext(), uri)
        if (IntentUtils.hasIntent(requireContext(), intent)) {
            takeCameraLauncher.launch(intent)
        }
    }

    private fun takePhoto() {
        autoBackgroundDetection = false
        val intent = IntentUtils.getImageGalleryIntent()
        getContentLauncher.launch(intent)
    }

    private val isActive: Boolean
        get() {
            val isDeactivated = isRemoving || isDetached || context == null
            return !isDeactivated
        }

    private fun loadUserProfileUrl(url: String) {
        if (isActive) {
            Glide.with(requireContext())
                .load(url)
                .circleCrop()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .error(
                    setTintList(
                        requireContext(),
                        R.drawable.icon_user,
                        SendbirdUIKit.getDefaultThemeMode().monoTintResId
                    )
                )
                .into(binding.ivProfileView)
        }
    }

    companion object {
        private val REQUIRED_PERMISSIONS = PermissionUtils.CAMERA_PERMISSION
        private fun getPermissionGuideMessage(context: Context, permission: String): String {
            val textResId: Int = if (Manifest.permission.CAMERA == permission) {
                com.sendbird.uikit.R.string.sb_text_need_to_allow_permission_camera
            } else {
                com.sendbird.uikit.R.string.sb_text_need_to_allow_permission_storage
            }
            return String.format(Locale.US, context.getString(textResId), ContextUtils.getApplicationName(context))
        }
    }
}
